core/shared/src/main/scala/scalaz/zio/Retry.scala

```diff
@@ -0,0 +1,428 @@
+
+  /**
+   * A retry strategy that always succeeds with the specified constant state.
+   */
+  final def point[E, S](s: => S): Retry[E, S] =
+    Retry[E, S](IO.point(s), (_, s) => IO.now(s))
+
+  /**
+   * A retry strategy that always succeeds, collecting all errors into a list.
+   */
+  final def errors[E]: Retry[E, List[E]] =
+    Retry[E, List[E]](IO.now(Nil), (e, l) => IO.now(e :: l))
+
+  /**
+   * A retry strategy that always retries and counts the number of retries.
+   */
+  final def counted[E]: Retry[E, Int] =
+    Retry[E, Int](IO.now(0), (_, i) => IO.now(i + 1))
+
+  /**
+   * A retry strategy that always retries and computes the time since the
+   * beginning of the process.
+   */
+  final def elapsed[E]: Retry[E, Duration] = {
+    val nanoTime = IO.sync(System.nanoTime())
+
+    Retry[E, (Long, Long)](nanoTime.seq(IO.now(0L)), (_, t) => nanoTime.map(t2 => (t._1, t2 - t._1)))
+      .map(t => Duration(t._2, TimeUnit.NANOSECONDS))
+  }
+
+  /**
+   * A retry strategy that will keep retrying until the specified number of
+   * retries is reached.
+   */
+  final def retries[E](max: Int): Retry[E, Int] = counted.untilState(_ >= max)
+
+  /**
+   * A retry strategy that will keep retrying until the specified duration has
+   * elapsed.
+   */
+  final def duration[E](duration: Duration): Retry[E, Duration] = {
+    val nanos = duration.toNanos
+
+    elapsed.untilState(_.toNanos >= nanos)
+  }
+
+  /**
+   * A retry strategy that will always succeed, waiting the specified fixed
+   * duration between attempts.
+   */
+  final def fixed[E](duration: Duration): Retry[E, Int] =
+    counted.updated((_, _, io) => io.delay(duration))
+
+  final def backoff[E](base: Duration,
+                       sleep: (Int, Duration) => IO[E, Duration],
+                       cap: Option[Duration] = None): Retry[E, (Int, Duration)] = {
+    val up: ((Int, Duration)) => IO[E, (Int, Duration)] = {
+      case (n, d) => sleep(n, d).map(d0 => (n + 1, d0)).delay(cap.fold(d)(_.min(d)))
+    }
+    Retry[E, (Int, Duration)](IO.now((0, base)), (_, s) => up(s))
+  }
+
+  /**
+   * A retry strategy that will always succeed, but will wait a certain amount
+   * between retries, given by `base * factor.pow(n)`, where `n` is the
+   * number of retries so far.
+   */
+  final def exponentialBackoff[E](base: Duration,
+                                  factor: Double = 2.0,
+                                  cap: Option[Duration] = None): Retry[E, (Int, Duration)] =
+    backoff(base, (n, _) => IO.now(base * math.pow(factor, n.doubleValue)), cap)
+
+  final def fullJitter[E](base: Duration,
+                          factor: Double = 2.0,
+                          cap: Option[Duration] = None): Retry[E, (Int, Duration)] = {
+    def jitter(n: Int) = {
+      val exp = base * math.pow(factor, n.doubleValue)
+      IO.sync(util.Random.nextDouble()).map(exp * _)
+    }
+    backoff(base, (n, _) => jitter(n), cap)
+  }
+
+  final def equalJitter[E](base: Duration,
```

Comment:
Both `jitter` and `fullJitter` have the same structure and we can unify them at the cost of another param.

Something else occurs to me: we would achieve greater composability with retry strategies that compute the next delay (in `S`), but do not actually sleep. Then after you're done modifying or jittering the delay, you could call `retry.sleep` (which would require that `S` == `Duration`).

Don't worry about that in this PR, I will play around with it when we get it merged.

Reply:
> Both jitter and fullJitter have the same structure and we can unify them at the cost of another param.

The param being a selector flag?

Reply:
I was thinking more numbers that would be multiplied by the various terms.